### PR TITLE
[master] add pchids not-exist-in-any-template in the response of edit_fcp_template API

### PR DESCRIPTION
**Pull request by wngzhe**
code change:
    change return of edit_fcp_template add orphan values in return value dict

result example:
 ```
{
              'fcp_template': {
                'name': 'bjcb-test-template',
                'id': '36439338-db14-11ec-bb41-0201018b1dd2',
                'description': 'This is Default template',
                'host_default': True,
                'storage_providers': ['sp4', 'v7k60'],
                'min_fcp_paths_count': 2,
                'pchids': {
                    'add' : ['C'],
                    'delete' : {
                        'all': ['D', 'E'],
                        'not_exist_in_any_template': ['F']
                    },
                    'all' : ['A', 'B', 'C']
                }
              }
            }
```

### CODE REVIEW

File: zvmsdk/database.py [hunk 1]
```diff
@@ -1436,11 +1444,19 @@ def edit_fcp_template(self, fcp_template_id, name=None, description=None,
...
                 'pchids': {
                     'add': add_items,
-                    'del': del_items,
-                    'all': all_items
+                    'delete': delete_dict,
+                    'all': final_phid_list
                 }
             }}
 
+    def get_all_pchids_used_by_template(self):
```
Comment (dongyanyang) [inline zvmsdk/database.py:1452]: 1. current loop will call several times of db query.
Please check the implementation of get_pchids_by_fcp_template, what you need is to just remove the condition of "where xxx" in the SQL cmd.

2. move the function to database.py, refer to get_pchids_by_fcp_template.

File: zvmsdk/database.py [hunk 2]
```diff
@@ -1424,7 +1429,10 @@ def edit_fcp_template(self, fcp_template_id, name=None, description=None,
             final_phid_list = self.get_pchids_by_fcp_template(fcp_template_id)
             add_items = list(set(final_phid_list) - set(ori_phid_list))
             del_items = list(set(ori_phid_list) - set(final_phid_list))
-            all_items = final_phid_list
+            all_items_in_db = self.get_all_pchids_in_db()
+            all_items_in_template = self.get_all_pchids_used_by_template()
+            orphan_items = list(set(all_items_in_db) - set(all_items_in_template))
```
Comment (dongyanyang) [inline zvmsdk/database.py:1434]: this calculation is wrong.
1. remove the function get_all_pchids_in_db(), it only represents all the pchids on this z/VM LPAR.
2. not_used_in_any_template = set(del_items) - set(all_pchids_used_by_templates)
3. do not call it orphan_items, rename it to not_used_in_any_template

**Review by dongyanyang — CHANGES_REQUESTED** [review]

File: zvmsdk/database.py [hunk 3]
```diff
@@ -1236,9 +1236,14 @@ def edit_fcp_template(self, fcp_template_id, name=None, description=None,
                 'host_default': True,
                 'storage_providers': ['sp4', 'v7k60'],
                 'min_fcp_paths_count': 2,
-                'pchids': {'add': [],
-                           'del': [],
-                           'all': ['0a20']},
+                'pchids': {
+                    'add' : ['C'],
+                    'delete' : {
+                        'all': [D, E],
```
Comment (jackydalong) [inline zvmsdk/database.py:1242]: change to 'all': ['D', 'E'],

File: zvmsdk/database.py [hunk 4]
```diff
@@ -1236,9 +1236,14 @@ def edit_fcp_template(self, fcp_template_id, name=None, description=None,
...
+                        'not_exist_in_any_template': [D]
```
Comment (jackydalong) [inline zvmsdk/database.py:1243]: change from [D] to ['D']

File: zvmsdk/database.py [hunk 5]
```diff
@@ -1236,9 +1236,14 @@ def edit_fcp_template(self, fcp_template_id, name=None, description=None,
...
+                         },
+                    'all' : ['A', 'B', 'C']
+                    }
```
Comment (jackydalong) [inline zvmsdk/database.py:1246]: Missing the following info
```python
        'cpc_sn': '23432',      <-------
        'cpc_name': 'M54',      <-------
        'lpar': 'ZVM4OCP1',      <------- Not z/VM name BOEM5401
        "hypervisor_hostname": "BOEM5401" <-------
```

File: zvmsdk/database.py [hunk 6]
```diff
@@ -1236,9 +1236,14 @@ def edit_fcp_template(self, fcp_template_id, name=None, description=None,
...
                 'min_fcp_paths_count': 2,
-                'pchids': {'add': [],
-                           'del': [],
-                           'all': ['0a20']},
+                'pchids': {
+                    'add' : ['C'],
+                    'delete' : {
+                        'all': [D, E],
+                        'not_exist_in_any_template': [D]
+                         },
```
Comment (jackydalong) [inline zvmsdk/database.py:1244]: delete 4 whitespace before }

File: zvmsdk/database.py (hunk 5 above)
Comment (jackydalong) [inline zvmsdk/database.py:1246]: delete 4 whitespace before }

File: zvmsdk/database.py [hunk 7]
```diff
@@ -1424,7 +1429,10 @@ def edit_fcp_template(self, fcp_template_id, name=None, description=None,
             final_phid_list = self.get_pchids_by_fcp_template(fcp_template_id)
```
Comment (jackydalong) [inline zvmsdk/database.py:1429]: There are many variable with typo `xx_phid_xx` introduced by your last PR, please rename all of them as `xx_pchid_xx`

**Comment by jackydalong** [comment]
suggest changing the title 
from 
```
change return of edit_fcp_template add orphan values in return value dict
```
to
```
add pchids-not-exist-in-any-template in the response of edit_fcp_template API
```

File: zvmsdk/database.py [hunk 8]
```diff
@@ -1424,7 +1429,10 @@ def edit_fcp_template(self, fcp_template_id, name=None, description=None,
...
             add_items = list(set(final_phid_list) - set(ori_phid_list))
             del_items = list(set(ori_phid_list) - set(final_phid_list))
-            all_items = final_phid_list
```
Comment (jackydalong) [inline zvmsdk/database.py:1427]: suggest change
from 
```
add_items, del_items, all_items
```
to
```
add_pchids, del_pchids, all_pchids
```

File: zvmsdk/database.py [hunk 9]
```diff
@@ -1424,7 +1429,10 @@ def edit_fcp_template(self, fcp_template_id, name=None, description=None,
...
+            all_items_in_db = self.get_all_pchids_in_db()
+            all_items_in_template = self.get_all_pchids_used_by_template()
+            orphan_items = list(set(all_items_in_db) - set(all_items_in_template))
+            delete_dict = dict(all=del_items, not_exist_in_any_template=orphan_items)
```
Comment (jackydalong) [inline zvmsdk/database.py:1435]: change from
```python
            all_items_in_db = self.get_all_pchids_in_db()
            all_items_in_template = self.get_all_pchids_used_by_template()
            orphan_items = list(set(all_items_in_db) - set(all_items_in_template))
            delete_dict = dict(all=del_items, not_exist_in_any_template=orphan_items)
```
to
```python
pchids_from_all_templates = self.get_pchids_from_all_fcp_templates()
pchids_not_used_in_any_template = list(set(del_pchids) - set(pchids_from_all_templates))
delete_dict = dict(all=del_pchids, not_exist_in_any_template=pchids_not_used_in_any_template)
```

File: zvmsdk/database.py [hunk 10]
```diff
@@ -1436,11 +1444,19 @@ def edit_fcp_template(self, fcp_template_id, name=None, description=None,
                 'min_fcp_paths_count': self.get_min_fcp_paths_count(fcp_template_id),
                 'pchids': {
                     'add': add_items,
```
Comment (jackydalong) [inline zvmsdk/database.py:1446]: change to **'add': add_pchids**

File: zvmsdk/database.py [hunk 11]
```diff
@@ -1436,11 +1444,19 @@ def edit_fcp_template(self, fcp_template_id, name=None, description=None,
...
-                    'del': del_items,
-                    'all': all_items
```
Comment (jackydalong) [inline zvmsdk/database.py:1440]: change to **'all': all_pchids**

File: zvmsdk/database.py (hunk 1 above)
Comment (jackydalong) [inline zvmsdk/database.py:1452]: change the function name as 
```python
def get_pchids_from_all_fcp_templates(self)
```

File: zvmsdk/database.py [hunk 12]
```diff
@@ -1489,6 +1505,23 @@ def get_pchids_by_fcp_template(self, fcp_template_id):
                 pchids.append(item['pchid'])
         return pchids
 
+    def get_all_pchids_in_db(self):
```
Comment (jackydalong) [inline zvmsdk/database.py:1508]: remove this function, no needed so far

File: zvmsdk/volumeop.py [hunk 13]
```diff
@@ -1104,9 +1104,14 @@ def edit_fcp_template(self, fcp_template_id, name=None,
...
-                           'del': [],
-                           'all': ['0a20']},
+                'pchids': {
+                    'add' : ['C'],
+                    'delete' : {
+                        'all': [D, E],
+                        'not_exist_in_any_template': [D]
+                         },
+                    'all' : ['A', 'B', 'C']
+                    }
```
Comment (jackydalong) [inline zvmsdk/volumeop.py:1114]: delete 4 whitespace before }

File: zvmsdk/volumeop.py [hunk 14]
```diff
@@ -1104,9 +1104,14 @@ def edit_fcp_template(self, fcp_template_id, name=None,
...
                 'min_fcp_paths_count': 2,
-                'pchids': {'add': [],
-                           'del': [],
-                           'all': ['0a20']},
+                'pchids': {
+                    'add' : ['C'],
+                    'delete' : {
+                        'all': [D, E],
+                        'not_exist_in_any_template': [D]
+                         },
```
Comment (jackydalong) [inline zvmsdk/volumeop.py:1112]: change to
```python
               'pchids': {
                    'add' : ['C'],
                    'delete' : {
                        'all': ['D', 'E'],                <----- 'D', 'E'
                        'not_exist_in_any_template': ['D']    <---- 'D'
                     },            <--- delete 4 whitespace before }


```

File: zvmsdk/database.py (hunk 1 above)
Comment (jackydalong) [reply zvmsdk/database.py:1452]: Please strictly follow the good practice of other functions to add **function-level comment** to 
- describe each param (if any)
- describe return value, including an example

Comment (jackydalong) [reply zvmsdk/database.py:1452]: As mentioned [here](https://github.com/openmainframeproject/feilong/pull/738/files#r1287854209) , from the performance view, suggest change to
```python
        pchids = []
        with get_fcp_conn() as conn:
            result = conn.execute(
                "SELECT DISTINCT fcp.pchid "
                "FROM template_fcp_mapping AS tf "
                "INNER JOIN fcp "
                "ON tf.fcp_id=fcp.fcp_id ")

            raw = result.fetchall()
            for item in raw:
                pchids.append(item['pchid'])
        return pchids
```

File: zvmsdk/database.py (hunk 10 above)
Comment (wngzhe) [reply zvmsdk/database.py:1446]: this comes from issue comment ,need we change this key?

**Review by jackydalong — APPROVED** [review]
lgtm

**Review by dongyanyang — APPROVED** [review]

**Review by bjhuangr — APPROVED** [review]

**Pull request merged**